README_cmake.md

```diff
@@ -304,17 +695,219 @@ and the `USE_RUNTIME` option is ignored. Other options work as expected.
+described above.
+
+The main drawback of this approach is that creating accurate IMPORTED targets is
+difficult since predicting the names and locations of your binaries across all
+possible platform and CMake project generators is difficult. In particular, it
+is hard to predict executable extensions in cross-OS builds.
+
+### Use an emulator or run on device
+
+The
+[`CMAKE_CROSSCOMPILING_EMULATOR`](https://cmake.org/cmake/help/latest/variable/CMAKE_CROSSCOMPILING_EMULATOR.html)
+variable allows one to specify a command _prefix_ to run a target-system binary
+on the host machine. One could set this to a custom shell script that uploads
+the generator executable, runs it on the device and copies back the results.
+
+### Bypass CMake
+
+The previous two options ensure that the targets generated by
+`add_halide_library` will be _normal_ static libraries. This approach does not
+use ExternalProject, but instead produces `IMPORTED` targets. The main drawback
+of `IMPORTED` targets is that they are considered second-class in CMake. In
+particular, they cannot be installed with the typical `install(TARGETS)`
+command. Instead, they must be installed using `install(FILES)` and the
+`$<TARGET_FILE:>` generator expression.
+
+# Contributing CMake code to Halide
+
+When contributing new CMake code to Halide, keep in mind that the minimum
+version is 3.16. Therefore, it is possible (and indeed required) to use modern
+CMake best practices.
+
+Like any large and complex system with a dedication to preserving backwards
+compatibility, CMake is difficult to learn and full of traps. While not
+comprehensive, the following serves as a guide for writing quality CMake code
+and outlines the code quality expectations we have as they apply to CMake.
+
+## General guidelines and best practices
+
+The following are some common mistakes that lead to subtly broken builds.
+
+- **Reading the build directory.** While setting up the build, the build
+  directory should be considered _write only_. Using the build directory as a
+  read/write temporary directory is acceptable as long as all temp files are
+  cleaned up by the end of configuration.
+- **Prefer generator expressions.** Declarative is better than imperative and
+  this is no exception. Conditionally adding to a target property can leak
+  unwanted details about the build environment into packages. Some information
+  is not accurate or available except via generator expressions, eg. the build
+  configuration.
+- **Using the wrong variable.** `CMAKE_SOURCE_DIR` doesn't always point to the
+  Halide source root. When someone uses Halide via FetchContent, it will point
+  to _their_ source root instead. The correct variable is `Halide_SOURCE_DIR`.
+  If you want to know if the compiler is MSVC, check it directly with the `MSVC`
+  variable; don't use `WIN32`. That will be wrong when compiling with clang on
+  Windows.
+- **Using directory properties.** Directory properties have vexing behavior and
```

Comment:
Maybe include examples of 'bad, don't do this' and 'do this instead'

Reply:
I tried to do that a bit with the tables in the following sections, but I could definitely flesh this out more.